Add BMT workflows and update functions to build Buildings module

yiyi1991 opened the pull request:
This is a draft PR to track bmt development (issue #431, and its parent issue #483) for the milestone "build the Buildings sector on a baseline scenario". 

Key functions that may need code check
- message_ix_models.model.buildings.build.**build_B**
- message_ix_models.model.buildings.build.**prepare_data_B**

When the buildings module is run as part of the bmt workflow, it calls ``build_B``. It loads inputs (prices, STURM outputs, static demand) specified in `context.buildings` or from the config file `data/bmt/config.yaml`, calls `prepare_data_B` to derive parameter data from the base scenario and those inputs, and then applies the buildings spec to the scenario. 

The `prepare_data_B` function produces demand and technology parameters (input, output, capacity factor, etc.) for buildings commodities and the residual AFOFIO representation; it can also add materials linkage when`with_materials` is true.

The two functions should be able to refactor and apply to message_ix_models.model.buildings.**main** and message_ix_models.model.buildings.build.**prepare_data**, so that the same functions apply to previous projects (e.g., NAVIGATE). Another PR will be opened for such a possible refactor.

## How to review

- Read the diff and note that the CI checks all pass.
- (Optional) Run ``mix-models bmt run --from="M cloned" "BM built"`` to see if it works on an R12 scenario. The function should work once the ``test_bmt.py`` passed, which includes ``test_build_B_runs_with_minimal_data`` and ``test_build_B_runs_with_materials``. 
- (Optional) Run ``mix-models bmt run --from="BM solved" "BM reported"`` to see if the report works. This calls the legacy report on message-data branch `bmt_dev`.
- The documentation makes sense.

## PR checklist
- [x] Continuous integration checks all ✅ 
- [x] Add or expand tests; coverage checks both ✅
- [x] Add, expand, or update documentation.
- [x] Update doc/whatsnew.

yiyi1991 committed:
Add BMT cli commands

yiyi1991 committed:
Add bmt workflow

yiyi1991 committed:
Add context-based material build

yiyi1991 committed:
Add context-based cli build

- build based on context.material
- default attributes saved in config.py and passed to context.material
- mix-models material cli options overwrites defaul attributes
- should be compatible for both cli build and workflow build
- commented out for now

yiyi1991 committed:
Add buildings config

yiyi1991 committed:
Add context-based build buildings

yiyi1991 committed:
Add a placeholder for add_dac to skip errors

yiyi1991 committed:
Update build buidlings for SSP baselines

yiyi1991 committed:
Apply pre-commit auto-fixes

yiyi1991 committed:
Add optional way to add afofi and update workflow

yiyi1991 committed:
Update context build material

yiyi1991 committed:
Assign useful energy and demand levels and additional nots

yiyi1991 committed:
Standardize how to treat material double-counting

yiyi1991 committed:
Add material subtraction vetting

yiyi1991 committed:
Update workflow for takeaway BM scenario

yiyi1991 committed:
Update bm_subtraction method

yiyi1991 committed:
Add build_PM function for power sector material intensity

yiyi1991 committed:
Update worklfow with PM in the last step

khaeru commented on message_ix_models/model/material/build.py:
```diff
@@ -273,3 +273,78 @@ def make_spec(regions: str, materials: str or None = SPEC_LIST) -> Spec:
     return s
+
+
+# same as build(), but context-based
```
Here and in .buildings.build, I think it would be preferable to adjust the build() or main() function to take a Context object, and then decide what to do/not do based on settings attached to it, like a .material.Config or .buildings.Config instance.

This way we could hopefully avoid duplicating methods, and reduce the risk that 2+ different 'build' methods diverge from one another.

khaeru replied:
(As always, happy to try to help make this change.)

khaeru replied:
The key transport lines are:
https://github.com/iiasa/message-ix-models/blob/b51d8f08fd999c50bb0d3a8c4ef10d0d030c4fab/message_ix_models/model/transport/build.py#L647-L652
and:
https://github.com/iiasa/message-ix-models/blob/b51d8f08fd999c50bb0d3a8c4ef10d0d030c4fab/message_ix_models/model/transport/build.py#L529-L545

The latter seems complicated, but it is basically trying to cover for different CLI, testing, and other uses of the code, and also allow to selectively override one or a few Config settings. For a minimal implementation, you could try something like:
```py
if "material" in context:
    # Existing config instance, don't do anything
    log.info(f"Use existing materials config: {context.material!r}")
else:
    # There's no materials Config instance, create a new one
    context["material"] = material.Config(**kwargs)
```

khaeru commented on message_ix_models/util/context.py:
```diff
@@ -459,3 +461,8 @@ def write_debug_archive(self) -> None:
         """
         self.core.write_debug_archive()
+
+    def print_contents(self):
```
Probably equivalent to:
```python
print(repr(context.asdict()))
```

yiyi1991 replied:
ah, thx

yiyi1991 committed:
Update BMT workflow

yiyi1991 committed:
Remove placeholders

yiyi1991 committed:
Deal with sp_el_RC_RT

yiyi1991 committed:
Revert prepare_data to NAVIGATE version (commit 630d110)

yiyi1991 committed:
Add commodity definition

yiyi1991 committed:
Update build

yiyi1991 committed:
Add prepare_data_B

yiyi1991 committed:
Update set yaml

yiyi1991 committed:
Remove ACT/CAP bounds of rc techs in 2020/2025

yiyi1991 committed:
Limit RT end-uses

yiyi1991 committed:
Add RT_elec_share_RC_max share mode constraint

yiyi1991 committed:
Update workflow for the current BM setting

yiyi1991 committed:
Add build_T and update BMT workflow

yiyi1991 committed:
Revert non-Buildings commits

yiyi1991 committed:
Update bmt config

yiyi1991 committed:
Update BMT workflow with config

yiyi1991 committed:
Update build buildings

yiyi1991 committed:
Add test

yiyi1991 committed:
Update build buildings to restrict required relations to existing ones

yiyi1991 committed:
Add doc

yiyi1991 committed:
Revise test_buildings so it checks either afofi or afofio

yiyi1991 committed:
Update tech number as sp_el_RC_RT is added

yiyi1991 committed:
Add additional test

yiyi1991 committed:
Specify dict to pass Python 3.14 env test

yiyi1991 committed:
Specify dict for other possible duplication of input_cap_new

yiyi1991 committed:
Specify dic with time year

macflo8 commented on message_ix_models/data/technology.yaml:
```diff
@@ -2896,6 +2896,14 @@ sp_el_RC:
   output: [rc_spec, useful]
 
+sp_el_RC_RT:
+  name: sp_el_RC
```
```suggestion
  name: sp_el_RC_RT
```

macflo8 commented on message_ix_models/tests/model/test_bmt.py:
```diff
@@ -0,0 +1,484 @@
+            "time": ["year", "year"],
+            "level": ["useful", "useful"],
+        }
+    )
```
```suggestion
    common_dims = dict(
        node="R12_AFR", unit="GWa", time="year", level="useful", year=[2020, 2030]
    )
    prices = pd.DataFrame({"commodity": commodities})
    sturm_r = make_df("demand", **common_dims, commodity="resid_heat_electr", value=1.0)
    sturm_c = make_df("demand", **common_dims, commodity="comm_heat_electr", value=0.5)
    demand_static = make_df(
        "demand", **common_dims, commodity=["afofio_spec", "afofio_therm"], value=0
    ).assign(year=[2020, 2020])
```

macflo8 reviewed (commented):
This looks good to me overall!
I left a few minor suggestions. Mainly formatting improvements and small refactorings to make some things more concise.

If there is any function/module that I should particularly look at (e.g.: regarding modelling methodology), please tag me there.

yiyi1991 committed:
Apply suggestions from code review

Co-authored-by: Florian Maczek <62338954+macflo8@users.noreply.github.com>

khaeru committed:
Adjust import in .test_bmt

Address 3 F821 errors from ruff.

khaeru commented:
Rebased on current `main`.

khaeru commented on doc/bmt/index.rst:
Newly added documentation files need to be added to 1 of the existing `.. toctree::` directives somewhere, else they will not appear in the built docs.

I've done this in the commits added to the branch.

khaeru committed:
Move .buildings.Config to a dedicated submodule

Mirror structure in .model.{material,transport}.

khaeru committed:
Add ConfigHelper.read_file(…, key=…) argument

khaeru committed:
Subclass ConfigHelper for .buildings.Config

This allows the utility methods of the former class to be used on
instances of the latter.

khaeru committed:
Add .buildings.Config.method; METHOD enumeration

khaeru committed:
Add .buildings.Config.data_paths and defaults

khaeru committed:
Consolidate .buildings.build.{build_B,main}

- Rename prepare_method() → prepare_method_A() to mirror
  prepare_method_B().
- Switch logic in main() based on Config.method.
- Adjust imports of build_B() in other modules.
- Adjust type hints.

khaeru committed:
Simplify .bmt.config

- Remove all but load_buildings_config() function.
- Construct and return an instance of .buildings.Config().
- Adjust data/bmt/config.yaml to match.
- Adjust tests.

khaeru committed:
Format and document .bmt.workflow.generate()

Adjust code for single-line statements.

khaeru committed:
Format and document .bmt.__init__

khaeru committed:
Add and use get_spec(…, filter_relations=…) param

khaeru committed:
Add #433 to docs, whatsnew

- Move doc/bmt.rst → doc/api/model-bmt.rst.
- Add to toctree directives.
- Add recursive autodoc of module contents.
- Apply semantic line breaks.

khaeru committed:
Move buildings tests to tests.model.test_buildings

khaeru reviewed (approved):
Approved, pending checks passing. I've made a number of refinements on recent commits on this branch, will edit this comment to explain.

- Instead of a separate class `.bmt.config.BuildingsConfig`, expand `.buildings.config.Config` to carry the additional information needed.
  - This avoids the need for lots of new code to read files, set defaults, update, etc., since `.buildings.config.Config` already has those.
- `.buildings.build_B()` was a duplicate of `.buildings.main()`, with substantial overlap in code but a different name. Instead of such a duplicate:
  - Use the same function for both 'methods'.
  - Add a setting (`.buildings.Config.method`) to control which of the methods is used; this setting enters the function through an argument (`context`).
  - Use `if:`-blocks in the function to guard steps that happen in one method but not the other.
- Instead of calling `get_spec()` and then following with code to modify its return value, pass an argument that signals it to return something different, as desired.
- The organization of test modules should match the organization of the code being tested. So, for instance, the added `.model.buildings.build.prepare_data_B` should be tested in `.tests.model.test_buildings`. Tests for `.model.bmt` code should be in `.tests.model.test_bmt`.

khaeru commented on message_ix_models/cli.py:
```diff
@@ -172,6 +172,7 @@ def _log_threads(k: int, n: int):
     "message_ix_models.testing.cli",
     "message_ix_models.util.pooch",
     "message_ix_models.util.slurm",
+    "message_ix_models.model.bmt.cli",
```
Note to self: in a follow-up PR, ensure this list is sorted.

khaeru replied:
Resolved in #486.

khaeru commented on message_ix_models/model/buildings/build.py:
```diff
@@ -602,6 +638,249 @@ def prepare_data(
     return result
 
 
+def prepare_data_B(  # noqa: C901
```
Note to self: complexity here is 27.

In a follow up PR:
- At minimum, add a comment and mention in pyproject.toml, similar to other complex methods.
- Try to reduce the complexity somewhat.

khaeru replied:
> - At minimum, add a comment and mention in pyproject.toml, similar to other complex methods.

This is done in #486.

khaeru commented on message_ix_models/model/bmt/utils.py:
```diff
@@ -0,0 +1,288 @@
+    return mat_demand
+
+
+def build_PM(context, scenario: "Scenario", **kwargs) -> "Scenario":
```
This seems to be used directly as a workflow step in the function `.workflow.generate()`. A loose convention we have is that functions that are workflow steps should live in the same .workflow module where they are used, so this could be shifted.
